src/EdgeEval/updater.rs

```diff
@@ -1,52 +1,70 @@
+pub struct CompiledUpdate {
+    pub clock_index: u32,
+    pub value: i32,
+}
+
+impl fmt::Display for CompiledUpdate {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_fmt(format_args!(
+            "{}:={}",
+            format!("c:{}", self.clock_index).to_string().magenta(),
+            self.value
+        ))?;
+        Ok(())
+    }
+}
 
-/// Used to handle update expressions on edges
-pub fn updater(
-    updates: &[parse_edge::Update],
-    decl: &component::Declarations, //Will eventually be mutable
-    zone: &mut Federation,
-) {
-    for update in updates {
+impl CompiledUpdate {
+    pub fn compile(update: &parse_edge::Update, decl: &Declarations) -> CompiledUpdate {
         match update.get_expression() {
             BoolExpression::Int(val) => {
                 if let Some(&clock_index) = decl.get_clock_index_by_name(update.get_variable_name())
                 {
-                    zone.update(clock_index, *val);
+                    CompiledUpdate {
+                        clock_index,
+                        value: *val,
+                    }
                 } else {
-                    panic!("Attempting to update a clock which is not initialized")
+                    panic!(
+                        "Attempting to compile an update with a clock \"{}\" which is not in decl",
+                        update.get_variable_name()
+                    )
                 }
             }
             _ => {
                 panic!("Should not be able to assign to {:?} in update", update)
             }
         }
     }
-}
 
-/// Used to handle update expressions on edges
-pub fn state_updater(
-    updates: &[parse_edge::Update],
-    state: &mut component::State,
-    comp_index: usize,
-) {
-    for update in updates {
-        match update.get_expression() {
-            BoolExpression::Int(val) => {
-                if let Some(&clock_index) = state
-                    .get_declarations(comp_index)
-                    .get_clock_index_by_name(update.get_variable_name())
-                {
-                    state.zone.update(clock_index, *val);
-                } else {
-                    panic!("Attempting to update a clock which is not initialized")
-                }
-            }
-            _ => {
-                panic!("Should not be able to assign to {:?} in update", update)
-            }
+    pub fn apply(&self, zone: &mut Federation) {
+        zone.update(self.clock_index, self.value);
+    }
+
+    pub fn as_update(&self, clocks: &HashMap<String, u32>) -> parse_edge::Update {
+        let map: HashMap<u32, String> = clocks.clone().into_iter().map(|(l, r)| (r, l)).collect();
+
+        parse_edge::Update {
+            variable: map.get(&self.clock_index).unwrap().clone(),
+            expression: BoolExpression::Int(self.value),
         }
     }
+
+    pub fn apply_as_free(&self, zone: &mut Federation) {
+        zone.free_clock(self.clock_index);
```

Comment:
apply_as_free does not assume that the update sets clocks to 0, applying as free is _(when used with apply_as_guard)_ the inverse of an update e.g. _the zone where after the update it would satisfy the invariant_.
The idea is to get the zone in the source location where if x were set to 5 it would satisfy the invariant  in the target location (in your example, x=5 as the invariant), so if the edge contained an `x:=5` update any value of x would satisfy the invariant, which is what the freeing of `x` does, removes all constraints on `x`. If the edge contained an `x:=2` update it would never be enabled which is why apply_as_free will almost always be used with apply_as_guard (freeing a clock on an empty federation remains an empty federation).